File: src/ObjectManager/Object.Tes/FilePacks/Records/345-APPA.Alchem Apparatus.cs
using OA.Core;

namespace OA.Tes.FilePacks.Records
{
    public class APPARecord : Record, IHaveEDID, IHaveMODL
    {
        // TESX
        public struct DATAField
        {
            public byte Type; // 0 = Mortar and Pestle, 1 = Albemic, 2 = Calcinator, 3 = Retort
            public float Quality;
            public float Weight;
            public int Value;

            public DATAField(UnityBinaryReader r, int dataSize, GameFormatId formatId)
            {
                if (formatId == GameFormatId.TES3)
                {
                    Type = (byte)r.ReadLEInt32();
                    Quality = r.ReadLESingle();
                    Weight = r.ReadLESingle();
                    Value = r.ReadLEInt32();
                    return;
                }
                Type = r.ReadByte();
                Value = r.ReadLEInt32();
                Weight = r.ReadLESingle();
                Quality = r.ReadLESingle();
            }
        }

        public override string ToString() => $"APPA: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public MODLGroup MODL { get; set; } // Model Name
        public STRVField FULL; // Item Name
        public DATAField DATA; // Alchemy Data
        public FILEField ICON; // Inventory Icon
        public FMIDField<SCPTRecord> SCRI; // Script Name

        public override bool CreateField(UnityBinaryReader r, GameFormatId formatId, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID":
                case "NAME": EDID = new STRVField(r, dataSize); return true;
                case "MODL": MODL = new MODLGroup(r, dataSize); return true;
                case "MODB": MODL.MODBField(r, dataSize); return true;
                case "MODT": MODL.MODTField(r, dataSize); return true;
                case "FULL":
                case "FNAM": FULL = new STRVField(r, dataSize); return true;
                case "DATA":
                case "AADT": DATA = new DATAField(r, dataSize, formatId); return true;
                case "ICON":
                case "ITEX": ICON = new FILEField(r, dataSize); return true;
                case "SCRI": SCRI = new FMIDField<SCPTRecord>(r, dataSize); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/345-APPA.Alchem Apparatus.swift
//
//  APPARecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class APPARecord: Record, IHaveEDID, IHaveMODL {
    // TESX
    public struct DATAField {
        public let type: UInt8 // 0 = Mortar and Pestle, 1 = Albemic, 2 = Calcinator, 3 = Retort
        public let quality: Float
        public let weight: Float
        public let value: Int32

        init(_ r: BinaryReader, _ dataSize: Int, _ format: GameFormatId) {
            guard format != .TES3 else {
                type = UInt8(r.readLEInt32())
                quality = r.readLESingle()
                weight = r.readLESingle()
                value = r.readLEInt32()
                return
            }
            type = r.readByte()
            value = r.readLEInt32()
            weight = r.readLESingle()
            quality = r.readLESingle()
        }
    }

    public override var description: String { return "APPA: \(EDID)" }
    public var EDID: STRVField = STRVField_empty  // Editor ID
    public var MODL: MODLGroup? = nil // Model Name
    public var FULL: STRVField! // Item Name
    public var DATA: DATAField! // Alchemy Data
    public var ICON: FILEField! // Inventory Icon
    public var SCRI: FMIDField<SCPTRecord>! // Script Name

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID",
             "NAME": EDID = r.readSTRV(dataSize)
        case "MODL": MODL = MODLGroup(r, dataSize)
        case "MODB": MODL!.MODBField(r, dataSize)
        case "MODT": MODL!.MODTField(r, dataSize)
        case "FULL",
             "FNAM": FULL = r.readSTRV(dataSize)
        case "DATA",
             "AADT": DATA = DATAField(r, dataSize, format)
        case "ICON",
             "ITEX": ICON = r.readSTRV(dataSize)
        case "SCRI": SCRI = FMIDField<SCPTRecord>(r, dataSize)
        default: return false
        }
        return true
    }
}
